Source/CarthageKit/Project.swift

```diff
@@ -453,23 +458,25 @@ public final class Project {
 						.flatMap(.concat, transform: unzip(archive:))
 						.flatMap(.concat) { directoryURL in
 							return frameworksInDirectory(directoryURL)
+								.flatMap(.merge) { url in
+									return checkFrameworkCompatibility(url)
+										.mapError { error in CarthageError.internalError(description: error.description) }
+								}
 								.flatMap(.merge, transform: self.copyFrameworkToBuildFolder)
 								.flatMap(.merge) { frameworkURL in
 									return self.copyDSYMToBuildFolderForFramework(frameworkURL, fromDirectoryURL: directoryURL)
 										.then(self.copyBCSymbolMapsToBuildFolderForFramework(frameworkURL, fromDirectoryURL: directoryURL))
 								}
 								.on(completed: {
 									_ = try? FileManager.default.trashItem(at: checkoutDirectoryURL, resultingItemURL: nil)
+								}, disposed: {
+									_ = try? FileManager.default.removeItem(at: directoryURL)
 								})
-								.then(SignalProducer<URL, CarthageError>(value: directoryURL))
 						}
-						.attemptMap { (temporaryDirectoryURL: URL) -> Result<Bool, CarthageError> in
-							do {
-								try FileManager.default.removeItem(at: temporaryDirectoryURL)
-								return .success(true)
-							} catch let error as NSError {
-								return .failure(.writeFailed(temporaryDirectoryURL, error))
-							}
+						.map { _ in true }
+						.flatMapError { error in
+							self._projectEventsObserver.send(value: .skippedInstallingBinaries(project: project, error: error))
+							return SignalProducer(value: false)
```

Comment:
The idea here is to handle any `CarthageError` produced by the `installBinaries` step. This is also why the `skippedInstallingBinaries` project message is generic, so that it can cover a range of errors--like errors in copying the frameworks artifacts. This is similar to the semantics for `skippedDownloadingBinaries` which is a general message that in practice is used only if you aren't authenticated to GitHub.

Also, I don't how we could move around the `flatMapError` and `mapError` without changing the pattern here at a larger scale. Since there are other failable steps in the process, those errors too have to be transforms into a `SignalProducer(value: false)`. Also, it seems DRYer to have a pattern of flowing errors through the observer chain and then handling the errors once per step, instead of introducing the pattern that errors should be handled as locally as possible. After all, none of the framework copying steps in here handle their errors locally, but instead pass the errors through the chain to be handled by the aforementioned `flatMapError` step to produce a `SignalProducer(value: false)`.